base/set.jl

```diff
@@ -661,9 +668,14 @@ Set([6, 12])
 ```
 """
 function replace!(new::Callable, A::Union{AbstractArray,AbstractDict,AbstractSet};
-                  count::Integer=typemax(Int))
-    count < 0 && throw(DomainError(count, "`count` must not be negative"))
-    count != 0 && _replace!(new, A, min(count, typemax(Int)) % Int)
+                  count::Union{Integer,Nothing}=nothing)
```

Comment:
Here we are using the "new API" to implement the functionality for certain types, so it's fine to use `replace!`, but then again I think we should keep calling `_replace!` the specific implementation methods, which have the signature `_replace!(new, res, A, count)`, which really is an implementation detail and is not part of any API.

Reply:
The problem with using `_replace!` is that custom types would have either to override `_replace!` (which is unexported and doesn't have a very appealing name), or to reimplement all `replace` and `replace!`methods directly. Here that implies duplicating some boilerplate (with some tricks, like checking `count === nothing` and `min(count, typemax(Int)) % Int)`, which one can easily get wrong). In other places, it even implies duplicating the call to `subtract_singletontype`, which is an unexported function and which people would better not have to bother with at all.

So I think it's much better to recommend package authors to implement a single low-level method, be it `replace!` or another unexported method (ideally with a better name than `_replace!`).